google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java

```diff
@@ -619,42 +628,63 @@ public URL signUrl(BlobInfo blobInfo, long duration, TimeUnit unit, SignUrlOptio
+        isV4
+            ? TimeUnit.SECONDS.convert(unit.toMillis(duration), TimeUnit.MILLISECONDS)
+            : TimeUnit.SECONDS.convert(
+                getOptions().getClock().millisTime() + unit.toMillis(duration),
+                TimeUnit.MILLISECONDS);
 
     StringBuilder stPath = new StringBuilder();
     if (!blobInfo.getBucket().startsWith(PATH_DELIMITER)) {
       stPath.append(PATH_DELIMITER);
     }
     stPath.append(blobInfo.getBucket());
-    if (!blobInfo.getBucket().endsWith(PATH_DELIMITER)) {
+    if (!blobInfo.getBucket().endsWith(PATH_DELIMITER)
+        && !Strings.isNullOrEmpty(blobInfo.getName())) {
       stPath.append(PATH_DELIMITER);
     }
     if (blobInfo.getName().startsWith(PATH_DELIMITER)) {
       stPath.setLength(stPath.length() - 1);
     }
 
     String escapedName = UrlEscapers.urlFragmentEscaper().escape(blobInfo.getName());
-    stPath.append(escapedName.replace("?", "%3F"));
+    stPath.append(escapedName.replace("?", "%3F").replace(";", "%3B"));
 
     URI path = URI.create(stPath.toString());
 
     try {
-      SignatureInfo signatureInfo = buildSignatureInfo(optionMap, blobInfo, expiration, path);
-      byte[] signatureBytes =
-          credentials.sign(signatureInfo.constructUnsignedPayload().getBytes(UTF_8));
+      SignatureInfo signatureInfo =
+          buildSignatureInfo(optionMap, blobInfo, expiration, path, credentials.getAccount());
+      String unsignedPayload = signatureInfo.constructUnsignedPayload();
+      byte[] signatureBytes = credentials.sign(unsignedPayload.getBytes(UTF_8));
       StringBuilder stBuilder = new StringBuilder();
       if (optionMap.get(SignUrlOption.Option.HOST_NAME) == null) {
         stBuilder.append(STORAGE_XML_HOST_NAME).append(path);
       } else {
         stBuilder.append(optionMap.get(SignUrlOption.Option.HOST_NAME)).append(path);
       }
 
-      String signature =
-          URLEncoder.encode(BaseEncoding.base64().encode(signatureBytes), UTF_8.name());
-      stBuilder.append("?GoogleAccessId=").append(credentials.getAccount());
-      stBuilder.append("&Expires=").append(expiration);
-      stBuilder.append("&Signature=").append(signature);
+      if (isV4) {
+        BaseEncoding encoding = BaseEncoding.base16().lowerCase();
+        String signature = URLEncoder.encode(encoding.encode(signatureBytes), UTF_8.name());
+        stBuilder.append("?");
+        stBuilder.append(signatureInfo.constructV4QueryString());
+        stBuilder.append("&X-Goog-Signature=").append(signature);
+      } else {
+        BaseEncoding encoding = BaseEncoding.base64();
+        String signature = URLEncoder.encode(encoding.encode(signatureBytes), UTF_8.name());
+        stBuilder.append("?");
+        stBuilder.append("GoogleAccessId=").append(credentials.getAccount());
+        stBuilder.append("&Expires=").append(expiration);
+        stBuilder.append("&Signature=").append(signature);
+
+        log.warning(
```

Comment:
One last comment: Jesse is it possible to only log this when they use signedUrl? Possibly using a third enum to determine if it's using v2 by default or if the user wants to use v2.